libraries/AP_HAL_PX4/CAN.cpp

```diff
@@ -0,0 +1,1035 @@
+{
+    uavcan::CanSelectMasks msk;
+
+    if (can_number_ >= 1) {
+        // Iface 0
+        msk.read = if0_.isRxBufferEmpty() ? 0 : 1;
+
+        if (pending_tx[0] != UAVCAN_NULLPTR) {
+            msk.write = if0_.canAcceptNewTxFrame(*pending_tx[0]) ? 1 : 0;
+        }
+    }
+
+    if (can_number_ >= 2) {
+        if (!if1_.isRxBufferEmpty()) {
+            msk.read |= 1 << 1;
+        }
+
+        // Iface 1
+        if (pending_tx[1] != UAVCAN_NULLPTR) {
+            if (if1_.canAcceptNewTxFrame(*pending_tx[1])) {
+                msk.write |= 1 << 1;
+            }
+        }
+    }
+
+    return msk;
+}
+
+bool PX4CANDriver::hasReadableInterfaces() const
+{
+    if (can_number_ >= 2) {
+        return !if0_.isRxBufferEmpty() || !if1_.isRxBufferEmpty();
+    }
+
+    return !if0_.isRxBufferEmpty();
+}
+
+int16_t PX4CANDriver::select(uavcan::CanSelectMasks& inout_masks,
+                             const uavcan::CanFrame* (&pending_tx)[uavcan::MaxCanIfaces], const uavcan::MonotonicTime blocking_deadline)
+{
+    const uavcan::CanSelectMasks in_masks = inout_masks;
+    const uavcan::MonotonicTime time = clock::getMonotonic();
+
+    if (can_number_ >= 1) {
+        if0_.discardTimedOutTxMailboxes(time); // Check TX timeouts - this may release some TX slots
+        {
+            CriticalSectionLocker cs_locker;
+            if0_.pollErrorFlagsFromISR();
+        }
+    }
+
+    if (can_number_ >= 2) {
+#if UAVCAN_STM32_NUM_IFACES > 1
+        if1_.discardTimedOutTxMailboxes(time);
+        {
+            CriticalSectionLocker cs_locker;
+            if1_.pollErrorFlagsFromISR();
+        }
+#endif
+    }
+
+    inout_masks = makeSelectMasks(pending_tx); // Check if we already have some of the requested events
+    if ((inout_masks.read & in_masks.read) != 0 || (inout_masks.write & in_masks.write) != 0) {
+        return 1;
+    }
+
+    (void) update_event_.wait(blocking_deadline - time); // Block until timeout expires or any iface updates
+    inout_masks = makeSelectMasks(pending_tx); // Return what we got even if none of the requested events are set
+    return 1; // Return value doesn't matter as long as it is non-negative
+}
+
+void PX4CANDriver::initOnce(uint8_t can_number)
+{
+    /*
+     * CAN1, CAN2
+     */
+    {
+        CriticalSectionLocker lock;
+        if (can_number >= 1) {
+            modifyreg32(STM32_RCC_APB1ENR, 0, RCC_APB1ENR_CAN1EN);
+        }
+        if (can_number >= 2) {
+            modifyreg32(STM32_RCC_APB1ENR, 0, RCC_APB1ENR_CAN2EN);
+        }
+    }
+
+    if (can_number >= 1) {
+#if defined(GPIO_CAN1_RX)
+        stm32_configgpio(GPIO_CAN1_RX);
+        stm32_configgpio(GPIO_CAN1_TX);
+#endif
+    }
+    if (can_number >= 2) {
+#if defined(GPIO_CAN2_RX)
+        stm32_configgpio(GPIO_CAN2_RX | GPIO_PULLUP);
+        stm32_configgpio(GPIO_CAN2_TX);
+#endif
+    }
+#if !defined(GPIO_CAN1_RX) &&  !defined(GPIO_CAN2_RX)
+# error  "Need to define GPIO_CAN1_RX and/or GPIO_CAN2_RX"
```

Comment:
Spearated. For PX4 both needs to be defined, even if not used.
TX does not matter as it's GPIO is set by hardware and we are not setting it.